opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/aggregation.py

```diff
@@ -349,6 +362,334 @@ def collect(
         return current_point
 
 
+# pylint: disable=protected-access
+class _ExponentialBucketHistogramAggregation(_Aggregation[HistogramPoint]):
+    # _min_max_size and _max_max_size are the smallest and largest values
+    # the max_size parameter may have, respectively.
+
+    # _min_max_size is is the smallest reasonable value which is small enough
+    # to contain the entire normal floating point range at the minimum scale.
+    _min_max_size = 2
+
+    # _max_max_size is an arbitrary limit meant to limit accidental creation of
+    # giant exponential bucket histograms.
+    _max_max_size = 16384
+
+    def __init__(
+        self,
+        attributes: Attributes,
+        start_time_unix_nano: int,
+        # This is the default maximum number of buckets per positive or
+        # negative number range.  The value 160 is specified by OpenTelemetry.
+        # See the derivation here:
+        # https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#exponential-bucket-histogram-aggregation)
+        max_size: int = 160,
+    ):
+        super().__init__(attributes)
+        # max_size is the maximum capacity of the positive and negative
+        # buckets.
+        if max_size < self._min_max_size:
+            raise Exception("size {max_size} is smaller than {min_max_size}")
+
+        if max_size > self._max_max_size:
+            raise Exception("size {max_size} is larter than {max_max_size}")
+
+        self._max_size = max_size
+
+        # _sum is the sum of all the values aggregated by this aggregator.
+        self._sum = 0
+
+        # _count is the count of all calls to aggregate.
+        self._count = 0
+
+        # _zero_count is the count of all the calls to aggregate when the value
```

Comment:
> Doesn't this include values near zero but not exactly to zero (something about subnormals?)

I think it does: `zero_count` is kept to populate [this](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/metrics/v1/metrics.proto#L523), described as:

>   // `zero_count` is the count of values that are either exactly zero or
  // within the region considered zero by the instrumentation at the
  // tolerated degree of precision.  This bucket stores values that
  // cannot be expressed using the standard exponential formula as
  // well as values that have been rounded to zero.

The smallest possible float point number that can be represented with IEEE 754 is a _denormal_ number, [this one](https://github.com/open-telemetry/opentelemetry-python/blob/main/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/ieee_754.md#normal-and-denormal-floating-point-numbers):

$$ 5 \times 10^{-324} $$

Assuming that the limit for _"the tolerated degree of precision for the instrumentation"_ (as mentioned above) is the smallest denormal value, then just comparing the value with 0 as we do [here](https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R453) should work, because Python treats values smaller than $5 \times 10^{-324}$ to be equal to $5 \times 10^{-324}$ or equal to zero:

``` python
Python 3.9.15 (main, Nov 28 2022, 20:14:06) 
Type 'copyright', 'credits' or 'license' for more information
IPython 8.7.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: 5e-324 == 0
Out[1]: False

In [2]: 5e-325 == 0
Out[2]: True

In [3]: 4e-325
Out[3]: 0.0

In [4]: 4e-324
Out[4]: 5e-324

In [5]: 3e-324
Out[5]: 5e-324

In [6]: 2e-324
Out[6]: 0.0

In [7]: 1e-324
Out[7]: 0.0

In [8]: 1e-324 == 0
Out[8]: True

In [9]: 4e-324 == 5e-324
Out[9]: True
```

Reply:
> https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/data-model.md#zero-count-and-zero-threshold
> 
> do we need to implement this? it looks like we are just checking if it is exactly zero

Right, we could implement `zero_threshold` too. Since that is an optional parameter we need a default behavior, I think the above one works for our purposes.

Reply:
> Assuming that the limit for "the tolerated degree of precision for the instrumentation" (as mentioned above) is the smallest denormal value

The mapping functions appear to map all subnormal values to a single index:

https://github.com/open-telemetry/opentelemetry-python/blob/edf4d6cd07b5ea922180b52c9e737b7be0d59f9f/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/exponent_mapping.py#L90-L91

https://github.com/open-telemetry/opentelemetry-python/blob/edf4d6cd07b5ea922180b52c9e737b7be0d59f9f/opentelemetry-sdk/src/opentelemetry/sdk/metrics/_internal/exponential_histogram/mapping/logarithm_mapping.py#L103-L105

Is that bucket not considered the zero count since all values less than it (including zero) are mapped to that index?

Reply:
> Is that bucket not considered the zero count since all values less than it (including zero) are mapped to that index?

Not including zero, `map_to_index` is not called if the value to aggregate is zero:

https://github.com/open-telemetry/opentelemetry-python/pull/2964/files#diff-f47777d9ac6e010c971c2252915bcbc8ec95dd61d24428c88d52b3f08a31f745R455

Reply:
Based on SIG discussion, I'm going to resolve this for now.